pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java

```diff
@@ -204,7 +223,26 @@ public double getDouble(int index) {
 
   @Override
   public String getUnpaddedString(int index, int numBytesPerValue, byte paddingByte, byte[] buffer) {
-    return StringUtil.decodeUtf8(getBytes(index, numBytesPerValue, buffer));
+    // Read the offset of the byte array first and then read the actual byte array.
+    int offset = _dataBuffer.getInt(_dataSectionStartOffSet + Integer.BYTES * index);
+
+    // To get the length of the byte array, we use the next byte array offset.
+    int length = _dataBuffer.getInt(_dataSectionStartOffSet + Integer.BYTES * (index + 1)) - offset;
+
+    byte[] b;
+    if (buffer != null && buffer.length >= length) {
+      b = buffer;
+    } else {
+      // Check if the current instance of ThreadLocal buffer is big enough. If not, resize it to double the size.
+      b = _reusableBytes.get();
+      if (b.length < length) {
+        b = new byte[nextPowerOf2(length)];
```

Comment:
Do we allow arbitary length strings? IIRC @Jackie-Jiang had put a max string length limit? If so, we could cap this allocation to that limit as well.

Reply:
max string is set in the schema and changeable.

Reply:
@mcvsubbu @mayankshriv Actually, good point. Jackie has merged the PR but this can be addressed in another PR too.

@mcvsubbu pointer to the config you're talking about please?

Reply:
IMO the current code is okay. Usually we put some power of 2 as the max length. I don't think it's worth nesting the reader with schema to save small memory.